src/main/java/org/jabref/model/openoffice/uno/UnoCrossRef.java

```diff
@@ -0,0 +1,75 @@
+package org.jabref.model.openoffice.uno;
+
+import com.sun.star.beans.PropertyVetoException;
+import com.sun.star.beans.UnknownPropertyException;
+import com.sun.star.beans.XPropertySet;
+import com.sun.star.lang.WrappedTargetException;
+import com.sun.star.lang.XMultiServiceFactory;
+import com.sun.star.text.ReferenceFieldPart;
+import com.sun.star.text.ReferenceFieldSource;
+import com.sun.star.text.XTextContent;
+import com.sun.star.text.XTextDocument;
+import com.sun.star.text.XTextRange;
+import com.sun.star.util.XRefreshable;
+
+public class UnoCrossRef {
+
+    private UnoCrossRef() { }
+
+    /**
+     * Update TextFields, etc. We use it to refresh cross-references in the document.
+     */
+    public static void refresh(XTextDocument doc) {
+        // Refresh the document
+        XRefreshable xRefresh = UnoCast.unoQI(XRefreshable.class, doc);
+        xRefresh.refresh();
+    }
+
+    /**
+     * Insert a clickable cross-reference to a reference mark,
+     * with a label containing the target's page number.
+     *
+     * May need a documentConnection.refresh() after, to update
+     * the text shown.
+     */
+    public static void insertReferenceToPageNumberOfReferenceMark(XTextDocument doc,
+                                                                  String referenceMarkName,
+                                                                  XTextRange cursor)
+        throws
+        CreationException,
+        UnknownPropertyException,
+        PropertyVetoException,
+        WrappedTargetException {
+
+        // based on: https://wiki.openoffice.org/wiki/Documentation/DevGuide/Text/Reference_Marks
+        XMultiServiceFactory msf = UnoCast.unoQI(XMultiServiceFactory.class, doc);
+        // Create a 'GetReference' text field to refer to the reference mark we just inserted,
+        // and get it's XPropertySet interface
+        XPropertySet xFieldProps;
+        try {
+            String name = "com.sun.star.text.textfield.GetReference";
+            xFieldProps = (XPropertySet) UnoCast.unoQI(XPropertySet.class,
+                                                       msf.createInstance(name));
+        } catch (Exception e) {
```

Comment:
Is it possible to catch a more specific Exception? You

Reply:
https://www.openoffice.org/api/docs/common/ref/com/sun/star/lang/XMultiServiceFactory.html#createInstance

- `raises(  ::com::sun::star::uno::Exception );`

https://www.openoffice.org/api/docs/common/ref/com/sun/star/uno/Exception.html

- All exceptions defined in UNO idl should derive from this exception.

https://wiki.openoffice.org/wiki/Documentation/DevGuide/ProUNO/Java/Mapping_of_Exception_Types

```
package com.sun.star.uno;
   public class Exception extends java.lang.Exception [...]
```

yes. `com.sun.star.uno.Exception` is not the same as `java,lang.Exception`
Modified. to `catch (con.sun.star.uno.Exception e)`